### PR TITLE
feat: workspace cache

### PR DESCRIPTION
**Pull request by dsherret**
This adds a workspace cache for use by the LSP, but also there's a bug fix for another edge case. I'll point out the test and code that fixes the edge case.

### CODE REVIEW

File: src/workspace/discovery.rs [hunk 1]
```diff
@@ -295,13 +305,56 @@ fn discover_workspace_config_files_for_single_dir(
...
+          // the workspace cache
+          let config_folder =
+            ConfigFolder::Single(DenoOrPkgJson::Deno(config_file));
+          return Ok(ConfigFileDiscovery::Workspace {
+            maybe_vendor_dir: resolve_vendor_dir(
+              config_folder.deno_json().map(|d| d.as_ref()),
+              opts.maybe_vendor_override.as_ref(),
+            ),
+            workspace: new_rc(Workspace::new_single(config_folder)),
+          });
```
Comment (dsherret) [inline src/workspace/discovery.rs:340]: This is a bug fix here. Basically, if someone specifies a path like `not-deno.json` then we shouldn't do workspace discovery because there's no way that this configuration file could be part of the workspace.

Additionally, if someone specifies a `deno.jsonc` file when there's a `deno.json` file, the `deno.json` file would be part of the workspace and not the `deno.jsonc`, so we should also not do workspace resolution in this case.

File: src/workspace/mod.rs [hunk 2]
```diff
@@ -4391,24 +4401,137 @@ mod test {
...
+          ..Default::default()
+        },
+      )
+      .unwrap_err();
+      assert_err(&err, &config_file_path);
+    }
+  }
+
+  #[test]
+  fn test_config_not_deno_workspace_member_non_natural_config_file_name() {
```
Comment (dsherret) [inline src/workspace/mod.rs:4453]: Test for bug fix: https://github.com/denoland/deno_config/pull/95/files#r1683560986

File: src/workspace/discovery.rs [hunk 3]
```diff
@@ -323,263 +376,63 @@ fn discover_workspace_config_files_for_single_dir(
...
       continue;
     };
-    if root_config_folder.is_workspace() {
+    if root_config_folder.has_workspace_members() {
       let maybe_vendor_dir = resolve_vendor_dir(
-        Some(&root_config_folder),
+        root_config_folder.deno_json().map(|d| d.as_ref()),
         opts.maybe_vendor_override.as_ref(),
       );
-      let mut final_members = BTreeMap::new();
```
Comment (dsherret) [inline src/workspace/discovery.rs:335]: Extracted out to function.

File: src/workspace/discovery.rs [hunk 4]
```diff
@@ -607,33 +464,290 @@ fn discover_workspace_config_files_for_single_dir(
...
+    }
+  } else {
+    let deno_json = maybe_deno_json?;
+    if deno_json.vendor() == Some(true) {
+      Some(deno_json.dir_path().join("vendor"))
+    } else {
+      None
+    }
   }
 }
```
Comment (dsherret) [inline src/workspace/discovery.rs:752]: Changed this to just accept a deno json instead of ConfigFolders which was needlessly complex

File: src/workspace/discovery.rs (hunk 1 above)
Comment (bartlomieju) [reply src/workspace/discovery.rs:340]: Wow, good catch 👍

**Review by bartlomieju — APPROVED** [review]
LGTM
